File: algorithms/solutions/week31/taurandat/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  int maxPathSum(TreeNode *root) {
    int maxPath = INT_MIN;
    dfsMaxPath(root, maxPath);
    return maxPath;
  }

  int dfsMaxPath(TreeNode *root, int &maxPath) {
    if (!root) return 0;
    int l = max(0, dfsMaxPath(root->left, maxPath));
    int r = max(0, dfsMaxPath(root->right, maxPath));
    maxPath = max(maxPath, l + r + root->val);
    return root->val + max(l, r);
  }
};

File: algorithms/solutions/week31/toanha/binary-tree-maximum-path-sum.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	res := math.MinInt64
	s := findMaxSum(root, &res)
	return max(res, s)
}

func findMaxSum(t *TreeNode, result *int) int {

	if t == nil {
		return 0
	}

	l := findMaxSum(t.Left, result)
	r := findMaxSum(t.Right, result)

	*result = max(*result, max(max(t.Val+l+r, t.Val+l), t.Val+r))

	return max(max(t.Val, t.Val+l), t.Val+r)
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func main() {

}
